From 83023c3eb40283a9467663aaad34aa51b38c97f7 Mon Sep 17 00:00:00 2001
From: Chris Williamson <chris@chrisaw.io>
Date: Mon, 16 Feb 2026 13:04:27 +0000
Subject: [PATCH] Add projector support to agold-cmd and include in GSI

Update agold-cmd to support ioctl and cmd subcommands for controlling
the Unihertz/Agold projector hardware. Add BnAgoldDaemonCallback for
receiving async results from the daemon cmd interface.

Include agold-cmd in PRODUCT_PACKAGES so it ships in the GSI image.
---
 base.mk            |  3 ++-
 cmds/agold-cmd.cpp | 30 +++++++++++++++++++-----------
 2 files changed, 21 insertions(+), 12 deletions(-)

diff --git a/base.mk b/base.mk
index 1928cdc..2f9a77a 100644
--- a/base.mk
+++ b/base.mk
@@ -147,7 +147,8 @@ PRODUCT_PACKAGES += \
 PRODUCT_PACKAGES += \
     lightsctl \
     lightsctl-aidl \
-    uevent
+    uevent \
+    agold-cmd
 
 PRODUCT_COPY_FILES += \
     device/phh/treble/files/adbd.rc:system/etc/init/adbd.rc
diff --git a/cmds/agold-cmd.cpp b/cmds/agold-cmd.cpp
index 9943557..462526a 100644
--- a/cmds/agold-cmd.cpp
+++ b/cmds/agold-cmd.cpp
@@ -1,22 +1,30 @@
 #include <iostream>
 #include <vendor/mediatek/hardware/agolddaemon/IAgoldDaemon.h>
+#include <vendor/mediatek/hardware/agolddaemon/BnAgoldDaemonCallback.h>
 #include <binder/IServiceManager.h>
+#include <utils/String8.h>
 
 using ::vendor::mediatek::hardware::agolddaemon::IAgoldDaemon;
+using ::vendor::mediatek::hardware::agolddaemon::BnAgoldDaemonCallback;
 using ::android::sp;
 
+class AgoldDaemonCallback : public BnAgoldDaemonCallback {
+    public:
+        ::android::binder::Status onResult(const ::android::String16& result) override {
+            std::cout << "Result: " << ::android::String8(result).c_str() << std::endl;
+            return ::android::binder::Status::ok();
+        }
+};
+
 int main(int argc, char **argv) {
 	auto svc = android::checkVintfService<IAgoldDaemon>();
-    android::String16 a;
-    svc->getNotGsi(&a);
-    std::cout << "notgsi is " << a << std::endl;
-    if (argc >= 2) {
-        android::String16 b(argv[1]);
-        svc->setNotGsi(b);
-    }
-    svc->getNotGsi(&a);
-    std::cout << "notgsi is " << a << std::endl;
-
-    svc->setCameraClientPackageName(android::String16("com.mediatek.camera"));
+    int32_t ret = 0;
 
+    if (std::string(argv[1]) == "ioctl") {
+        svc->SendMessageToIoctl(atoi(argv[2]), atoi(argv[3]), atoi(argv[4]), atoi(argv[5]), &ret);
+    } else {
+        sp<AgoldDaemonCallback> cb = new AgoldDaemonCallback();
+        svc->cmd(android::String16(argv[2]), cb);
+    }
+    printf("Returned %d\n", ret);
 }
-- 
2.52.0

